arrow-mtl/src/main/kotlin/arrow/mtl/extensions/statet.kt

```diff
@@ -294,3 +297,84 @@ interface StateTMonadTrans<S> : MonadTrans<Kind<ForStateT, S>> {
   override fun <G, A> Kind<G, A>.liftT(MG: Monad<G>): Kind2<Kind<ForStateT, S>, G, A> =
     StateT.liftF(MG, this)
 }
+
+@extension
+interface StateTMonadPlus<S, F> : MonadPlus<StateTPartialOf<S, F>>, StateTMonad<S, F>, StateTAlternative<S, F> {
+  override fun MF(): Monad<F>
+  override fun AF(): Alternative<F>
+}
+
+@extension
+interface StateTMonadLogic<S, F> : MonadLogic<StateTPartialOf<S, F>>, StateTMonadPlus<S, F> {
+  fun ML(): MonadLogic<F>
+  override fun MF(): Monad<F> = ML()
+  override fun AF(): Alternative<F> = ML()
+
+  override fun <A> Kind<StateTPartialOf<S, F>, A>.splitM(): Kind<StateTPartialOf<S, F>, Option<Tuple2<Kind<StateTPartialOf<S, F>, A>, A>>> =
+    this.fix().let { fa ->
+      StateT(
+        ML().run {
+          AndThen.id<S>().flatMap { s ->
+            AndThen(fa.runF).andThen {
+              it.splitM().flatMap { option ->
+                option.fold({ just(s toT Option.empty<Tuple2<Kind<StateTPartialOf<S, F>, A>, A>>()) }, { (fa, tupleSA) ->
+                  val (s1, a) = tupleSA
+                  just(s1 toT Option.just(StateT { _: S -> fa } toT a))
+                })
+              }
+            }
+          }
+        })
+    }
+
+  override fun <A> Kind<StateTPartialOf<S, F>, A>.interleave(other: Kind<StateTPartialOf<S, F>, A>): Kind<StateTPartialOf<S, F>, A> =
+    (this.fix() to other.fix()).let { (left, right) ->
+      StateT(
+        ML().run {
+          AndThen.id<S>().flatMap { s ->
+            AndThen(left.runF).andThen {
+              it.interleave(right.run(s))
+            }
+          }
```

Comment:
```suggestion
          AndThen.id<S>().flatMap { s ->
            AndThen(left.runF).flatMap { fa ->
              AndThen(right.runF).andThen {
                fa.interleave(it)
              }
            }
          }
```
Because it evaluates both strictly before running interleave anyway. Not sure if this suggestion thing works, but it shows the diff nicely ^^